crates/iceberg-catalog/src/catalog/s3_signer.rs

```diff
@@ -350,184 +348,109 @@ async fn validate_table_method<A: AuthZHandler>(
     Ok(())
 }
 
-const AWS_S3_ACCESS_POINTS: &[&str] = &["s3", "s3.dualstack", "s3-fips.dualstack", "s3-fips"];
-
 #[allow(clippy::too_many_lines)]
 fn validate_uri(
     // i.e. https://bucket.s3.region.amazonaws.com/key
-    request_uri: &url::Url,
+    parsed_url: &s3_utils::ParsedS3Url,
     // i.e. s3://bucket/key
     table_location: &str,
-    storage_profile: &S3Profile,
 ) -> Result<()> {
-    let table_location = url::Url::parse(table_location.trim_end_matches('/')).map_err(|e| {
-        ErrorModel::builder()
-            .code(http::StatusCode::INTERNAL_SERVER_ERROR.into())
-            .message("Failed to parse table location".to_string())
-            .r#type("FailedToParseTableLocation".to_string())
-            .source(Some(Box::new(e)))
-            .build()
-    })?;
-    let table_bucket = table_location.host_str().ok_or_else(|| {
-        ErrorModel::builder()
-            .code(http::StatusCode::INTERNAL_SERVER_ERROR.into())
-            .message("Table location does not have a bucket".to_string())
-            .r#type("TableLocationNoBucket".to_string())
+    let table_location = parse_s3_location(table_location)?;
+    let url_location = &parsed_url.location;
+
+    if !url_location.is_sublocation_of(&table_location) {
+        return Err(ErrorModel::builder()
+            .code(http::StatusCode::FORBIDDEN.into())
+            .message("Request URI does not match table location".to_string())
+            .r#type("RequestUriMismatch".to_string())
+            .stack(vec![
+                format!("Expected Location: {table_location}"),
+                format!("Actual Location: {url_location}"),
+            ])
             .build()
-    })?;
-    let table_key_virtual_host: Vec<_> = table_location
-        .path_segments()
-        .map(std::iter::Iterator::collect)
-        .ok_or(
-            ErrorModel::builder()
-                .code(http::StatusCode::INTERNAL_SERVER_ERROR.into())
-                .message("Table location does not have a key".to_string())
-                .r#type("TableLocationNoKey".to_string())
-                .build(),
-        )?;
-    let table_key_path_style = vec![table_bucket]
-        .into_iter()
-        .chain(table_key_virtual_host.clone())
-        .collect::<Vec<_>>();
-
-    let request_key: Vec<_> = request_uri
-        .path_segments()
-        .map(std::iter::Iterator::collect)
-        .unwrap_or_default();
+            .into());
+    }
+
+    Ok(())
+}
+
+pub(super) mod s3_utils {
+    use super::{ErrorModel, Result};
+    use crate::service::storage::S3Location;
+    use lazy_regex::regex;
+
+    #[derive(Debug)]
+    pub(super) struct ParsedS3Url {
+        pub(super) location: S3Location,
+        // Used endpoint without the bucket
+        #[allow(dead_code)]
```

Comment:
from september we can use this instead of allow dead code:

https://github.com/rust-lang/rust/pull/120924